File: CRUD/VB.Net/Aplicacion con SQL Server V2/Sistema/Sistema/Forms/FormProveedores.vb
' Written By Ismael Heredia in the year 2017

Imports Sistema.Aplicacion.Data

Public Class FormProveedores


    Dim proveedorDatos As New ProveedorDatos
    Dim funcion As New Funciones()
    Dim nuevo As Boolean = True

    ' Funciones

    Private Function validar() As Boolean

        If txtNombre.Text = "" Then
            MessageBox.Show("Ingrese el nombre")
            txtNombre.Focus()
            Return False
        End If

        If txtDireccion.Text = "" Then
            MessageBox.Show("Ingrese la direccion")
            txtDireccion.Focus()
            Return False
        End If

        If txtTelefono.Text = "" OrElse Not funcion.valid_number(txtTelefono.Text) Then
            MessageBox.Show("Ingrese el telefono")
            txtTelefono.Focus()
            Return False
        End If

        Return True

    End Function

    Private Sub cargarListaProveedores()
        lvProveedores.Items.Clear()
        Dim nombre_buscar As String = txtBuscar.Text
        Dim listaProveedores As List(Of Proveedor) = proveedorDatos.List(nombre_buscar)
        For Each proveedor As Proveedor In listaProveedores
            Dim item As New ListViewItem()
            item.Text = Convert.ToString(proveedor.nombre)
            item.SubItems.Add(Convert.ToString(proveedor.direccion))
            item.SubItems.Add(Convert.ToString(proveedor.telefono))
            item.SubItems.Add(Convert.ToString(proveedor.fecha_registro))
            item.Tag = Convert.ToString(proveedor.id)
            lvProveedores.Items.Add(item)
        Next
    End Sub

    Private Sub CargarCamposProveedor(id As Integer)
        Dim proveedor As Proveedor = proveedorDatos.Obtain(id)
        txtID.Text = Convert.ToString(proveedor.id)
        txtNombre.Text = proveedor.nombre
        txtDireccion.Text = proveedor.direccion
        txtTelefono.Text = proveedor.telefono.ToString()
    End Sub

    Private Sub limpiar()
        nuevo = True
        txtID.Text = ""
        txtNombre.Text = ""
        txtDireccion.Text = ""
        txtTelefono.Text = ""
    End Sub

    Private Sub btnBuscar_Click(sender As Object, e As EventArgs) Handles btnBuscar.Click
        recargarLista()
    End Sub

    Private Sub btnGrabar_Click(sender As Object, e As EventArgs) Handles btnGrabar.Click
        grabar()
    End Sub

    Private Sub FormProveedores_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        recargarLista()
    End Sub

    Private Sub lvProveedores_SelectedIndexChanged(sender As Object, e As EventArgs) Handles lvProveedores.SelectedIndexChanged
        If lvProveedores.SelectedIndices.Count > 0 AndAlso lvProveedores.SelectedIndices(0) <> -1 Then
            CargarCamposProveedor(Convert.ToInt32((lvProveedores.SelectedItems(0).Tag)))
        End If
    End Sub


    Private Sub grabar()
        If validar() Then

            Dim proveedor As New Proveedor()
            If txtID.Text <> "" Then
                proveedor.id = Convert.ToInt32(txtID.Text)
            End If
            proveedor.nombre = txtNombre.Text
            proveedor.direccion = txtDireccion.Text
            proveedor.telefono = Convert.ToInt32(txtTelefono.Text)
            proveedor.fecha_registro = funcion.fecha_del_dia()

            Dim grabar_ready As Boolean = False
            If nuevo Then
                If proveedorDatos.check_exists_proveedor_add(proveedor.nombre) Then
                    grabar_ready = False
                Else
                    grabar_ready = True
                End If
            Else
                If proveedorDatos.check_exists_proveedor_edit(proveedor.id,proveedor.nombre) Then
                    grabar_ready = False
                Else
                    grabar_ready = True
                End If
            End If

            If grabar_ready Then
                If nuevo Then
                    If (proveedorDatos.Add(proveedor)) Then
                        MessageBox.Show("Registro agregado")
                        status.Text = "[+] Registro agregado"
                        Me.Refresh()
                    Else
                        MessageBox.Show("Ha ocurrido un error en la base de datos")
                        status.Text = "[-] Ha ocurrido un error en la base de datos"
                        Me.Refresh()
                    End If
                Else
                    If (proveedorDatos.Update(proveedor)) Then
                        MessageBox.Show("Registro actualizado")
                        status.Text = "[+] Registro actualizado"
                        Me.Refresh()
                    Else
                        MessageBox.Show("Ha ocurrido un error en la base de datos")
                        status.Text = "[-] Ha ocurrido un error en la base de datos"
                        Me.Refresh()
                    End If
                End If
            Else
                MessageBox.Show("El proveedor " & proveedor.nombre & " ya existe")
                status.Text = "[-] El proveedor " & proveedor.nombre & " ya existe"
                Me.Refresh()
            End If

            cargarListaProveedores()
            limpiar()

        End If

    End Sub

    Private Sub agregar()
        status.Text = "[+] Programa en modo nuevo"
        Me.Refresh()
        nuevo = True
        limpiar()
        MessageBox.Show("Programa cargado en modo nuevo")
    End Sub


    Private Sub editar()
        status.Text = "[+] Programa en modo editar"
        Me.Refresh()
        nuevo = False
        MessageBox.Show("Programa cargado en modo editar")
    End Sub


    Private Sub borrar()

        Dim id As Integer = 0
        If lvProveedores.SelectedIndices.Count > 0 AndAlso lvProveedores.SelectedIndices(0) <> -1 Then
            id = Convert.ToInt32((lvProveedores.SelectedItems(0).Tag))
        End If

        Dim proveedor As Proveedor = proveedorDatos.Obtain(id)
        If MessageBox.Show("Seguro de borrar a " & proveedor.nombre, "Borrar registro", MessageBoxButtons.YesNo, MessageBoxIcon.Stop, MessageBoxDefaultButton.Button2) = DialogResult.Yes Then
            lvProveedores.Items.Clear()
            Dim sql As String = "delete from proveedores where id_proveedor=" & proveedor.id
            If proveedorDatos.Delete(proveedor) Then
                MessageBox.Show("Registro borrado")
                status.Text = "[+] Registro borrado"
                Me.Refresh()
            Else
                MessageBox.Show("Ha ocurrido un error en la base de datos")
                status.Text = "[-] Ha ocurrido un error en la base de datos"
                Me.Refresh()

            End If

            cargarListaProveedores()
            limpiar()

        End If

    End Sub


    Private Sub cancelar()
        status.Text = "[+] Programa cargado"
        Me.Refresh()
        nuevo = False
        limpiar()
        MessageBox.Show("Opcion cancelada")
    End Sub


    Private Sub recargarLista()
        cargarListaProveedores()
    End Sub


    Private Sub ItemAgregarProveedor_Click(sender As Object, e As EventArgs) Handles AgregarProveedorToolStripMenuItem.Click
        agregar()
    End Sub

    Private Sub ItemEditarProveedor_Click(sender As Object, e As EventArgs) Handles EditarProveedorToolStripMenuItem.Click
        editar()
    End Sub
    Private Sub ItemBorrarProveedor_Click(sender As Object, e As EventArgs) Handles BorrarProveedorToolStripMenuItem.Click
        borrar()
    End Sub

    Private Sub ItemCancelar_Click(sender As Object, e As EventArgs) Handles CancelarToolStripMenuItem.Click
        cancelar()
    End Sub

    Private Sub ItemRecargarLista_Click(sender As Object, e As EventArgs) Handles RecargarListaToolStripMenuItem.Click
        recargarLista()
    End Sub

    Private Sub ItemGrabar_Click(sender As Object, e As EventArgs) Handles GrabarToolStripMenuItem.Click
        grabar()
    End Sub


    Private Sub AgregarProveedorToolStripMenuItem1_Click(sender As Object, e As EventArgs) Handles AgregarProveedorToolStripMenuItem1.Click
        agregar()
    End Sub

    Private Sub EditarProveedorToolStripMenuItem1_Click(sender As Object, e As EventArgs) Handles EditarProveedorToolStripMenuItem1.Click
        editar()
    End Sub

    Private Sub BorrarProveedorToolStripMenuItem1_Click(sender As Object, e As EventArgs) Handles BorrarProveedorToolStripMenuItem1.Click
        borrar()
    End Sub

    Private Sub CancelarToolStripMenuItem1_Click(sender As Object, e As EventArgs) Handles CancelarToolStripMenuItem1.Click
        cancelar()
    End Sub

    Private Sub RecargarListaToolStripMenuItem1_Click(sender As Object, e As EventArgs) Handles RecargarListaToolStripMenuItem1.Click
        recargarLista()
    End Sub

    Private Sub GrabarToolStripMenuItem1_Click(sender As Object, e As EventArgs) Handles GrabarToolStripMenuItem1.Click
        grabar()
    End Sub

End Class

File: CRUD/C#/Aplicacion con SQL Server V2/Aplicacion/Forms/FormProveedores.cs
// Written By Ismael Heredia in the year 2017

using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Aplicacion.Models;
using Aplicacion.Data;

namespace aplicacion
{
    public partial class FormProveedores : Form
    {
        ProveedorDatos proveedorDatos = new ProveedorDatos();
        Funciones funcion = new Funciones();
        bool nuevo = true;

        public FormProveedores()
        {
            InitializeComponent();
        }

        // Funciones

        private bool validar()
        {

            if (txtNombre.Text == "")
            {
                MessageBox.Show("Ingrese el nombre");
                txtNombre.Focus();
                return false;
            }

            if (txtDireccion.Text == "")
            {
                MessageBox.Show("Ingrese la direccion");
                txtDireccion.Focus();
                return false;
            }

            if (txtTelefono.Text == "" || !funcion.valid_number(txtTelefono.Text))
            {
                MessageBox.Show("Ingrese el telefono");
                txtTelefono.Focus();
                return false;
            }

            return true;
        }

        private void cargarListaProveedores()
        {

            lvProveedores.Items.Clear();

            string nombre_buscar = txtBuscar.Text;

            List<Proveedor> listaProveedores = proveedorDatos.List(nombre_buscar);

            foreach(Proveedor proveedor in listaProveedores)
            {
                    ListViewItem item = new ListViewItem();
                    item.Text = Convert.ToString(proveedor.nombre);
                    item.SubItems.Add(Convert.ToString(proveedor.direccion));
                    item.SubItems.Add(Convert.ToString(proveedor.telefono));
                    item.SubItems.Add(Convert.ToString(proveedor.fecha_registro));
                    item.Tag = Convert.ToString(proveedor.id);
                    lvProveedores.Items.Add(item);
            }
        }

        private void CargarCamposProveedor(int id)
        {
            Proveedor proveedor = proveedorDatos.Get(id);
            txtID.Text = Convert.ToString(proveedor.id);
            txtNombre.Text = proveedor.nombre;
            txtDireccion.Text = proveedor.direccion;
            txtTelefono.Text = proveedor.telefono.ToString();
        }

        private void limpiar()
        {
            nuevo = true;
            txtID.Text = "";
            txtNombre.Text = "";
            txtDireccion.Text = "";
            txtTelefono.Text = "";
        }

        private void FormProveedores_Load(object sender, EventArgs e)
        {
            cargarListaProveedores();
        }

        private void lvProveedores_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (lvProveedores.SelectedIndices.Count > 0 && lvProveedores.SelectedIndices[0] != -1)
            {
                CargarCamposProveedor(Convert.ToInt32((lvProveedores.SelectedItems[0].Tag)));
            }
        }

        private void grabar()
        {
            if (validar())
            {

                Proveedor proveedor = new Proveedor();
                if (txtID.Text != "")
                {
                    proveedor.id = Convert.ToInt32(txtID.Text);
                }
                proveedor.nombre = txtNombre.Text;
                proveedor.direccion = txtDireccion.Text;
                proveedor.telefono = Convert.ToInt32(txtTelefono.Text);
                proveedor.fecha_registro = funcion.fecha_del_dia();

                bool grabar_ready = false;

                if (nuevo)
                {
                    if (proveedorDatos.check_exists_proveedor_add(proveedor.nombre))
                    {
                        grabar_ready = false;
                    }
                    else
                    {
                        grabar_ready = true;
                    }
                }
                else
                {
                    if (proveedorDatos.check_exists_proveedor_edit(proveedor.id,proveedor.nombre))
                    {
                        grabar_ready = false;
                    }
                    else
                    {
                        grabar_ready = true;
                    }
                }

                if (grabar_ready)
                {
                    if (nuevo)
                    {
                        if (proveedorDatos.Add(proveedor))
                        {
                            MessageBox.Show("Registro agregado");
                            tsStatus.Text = "[+] Registro agregado";
                            this.Refresh();
                        }

                        else
                        {
                            MessageBox.Show("Ha ocurrido un error en la base de datos");
                            tsStatus.Text = "[-] Ha ocurrido un error en la base de datos";
                            this.Refresh();
                        }
                    }
                    else
                    {
                        if (proveedorDatos.Update(proveedor))
                        {
                            MessageBox.Show("Registro actualizado");
                            tsStatus.Text = "[+] Registro actualizado";
                            this.Refresh();
                        }
                        else
                        {
                            MessageBox.Show("Ha ocurrido un error en la base de datos");
                            tsStatus.Text = "[-] Ha ocurrido un error en la base de datos";
                            this.Refresh();
                        }
                    }
                }
                else
                {
                    MessageBox.Show("El proveedor " + proveedor.nombre + " ya existe");
                    tsStatus.Text = "[-] El proveedor " + proveedor.nombre + " ya existe";
                    this.Refresh();
                }

                cargarListaProveedores();
                limpiar();

            }
        }

        private void agregar()
        {
            tsStatus.Text = "[+] Programa en modo nuevo";
            this.Refresh();
            nuevo = true;
            limpiar();
            MessageBox.Show("Programa cargado en modo nuevo");

        }

        private void editar()
        {
            tsStatus.Text = "[+] Programa en modo editar";
            this.Refresh();
            nuevo = false;
            MessageBox.Show("Programa cargado en modo editar");
        }

        private void borrar()
        {

            int id = 0;

            if (lvProveedores.SelectedIndices.Count > 0 && lvProveedores.SelectedIndices[0] != -1)
            {
                id = Convert.ToInt32((lvProveedores.SelectedItems[0].Tag));
            }

            Proveedor proveedor = proveedorDatos.Get(id);

            if (MessageBox.Show("Seguro de borrar a " + proveedor.nombre, "Borrar registro", MessageBoxButtons.YesNo, MessageBoxIcon.Stop, MessageBoxDefaultButton.Button2) == DialogResult.Yes)
            {
                lvProveedores.Items.Clear();

                if (proveedorDatos.Delete(proveedor))
                {
                    MessageBox.Show("Registro borrado");
                    tsStatus.Text = "[+] Registro borrado";
                    this.Refresh();

                }
                else
                {
                    MessageBox.Show("Ha ocurrido un error en la base de datos");
                    tsStatus.Text = "[-] Ha ocurrido un error en la base de datos";
                    this.Refresh();
                }

                cargarListaProveedores();
                limpiar();
  
            }
            
        }

        private void cancelar()
        {
            tsStatus.Text = "[+] Programa cargado";
            this.Refresh();
            nuevo = false;
            limpiar();
            MessageBox.Show("Opcion cancelada");
        }

        private void recargarLista()
        {
            cargarListaProveedores();
        }

        private void agregarProveedorToolStripMenuItem_Click(object sender, EventArgs e)
        {
            agregar();
        }

        private void editarProveedorToolStripMenuItem_Click(object sender, EventArgs e)
        {
            editar();
        }

        private void btnGrabar_Click(object sender, EventArgs e)
        {
            grabar();
        }

        private void borrarProveedorToolStripMenuItem_Click(object sender, EventArgs e)
        {
            borrar();
        }

        private void agregarProveedorToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            agregar();
        }

        private void editarProveedorToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            editar();
        }

        private void borrarProveedorToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            borrar();
        }

        private void cancelarToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            cancelar();
        }

        private void recargarListaToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            recargarLista();
        }

        private void cancelarToolStripMenuItem_Click(object sender, EventArgs e)
        {
            cancelar();
        }

        private void recargarListaToolStripMenuItem_Click(object sender, EventArgs e)
        {
            recargarLista();
        }

        private void grabarToolStripMenuItem_Click(object sender, EventArgs e)
        {
            grabar();
        }

        private void grabarToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            grabar();
        }

        private void btnBuscar_Click(object sender, EventArgs e)
        {
            recargarLista();
        }

    }
}
